### PR TITLE
Move external python glue logic into core

### PR DESCRIPTION
**Pull request by scheibelp**
This originally lived in the `python` build system and in this PR moves into core

This was originally plugged in generically: any extendable package could implement `_update_external_dependencies` and potentially resolve analogous issues. No other package in builtin did this (but a if a private repo took advantage of this, this PR would break them).

The changes are nearly a verbatim relocation (although one edge case for the old concretizer is removed), plus some renaming of methods to reflect the fact that it's just for python and not generic.

This logic is already tested (`spack unit-test -k external_python`).

### CODE REVIEW

File: lib/spack/spack/solver/asp.py [hunk 1]
```diff
@@ -4186,6 +4187,97 @@ def execute_explicit_splices(self):
...
+    If another package in the DAG depends on python, we use that
+    python for the dependency of the external. If not, we assume
+    that the external PythonPackage is installed into the same
+    directory as the python it depends on.
+    """
+    # TODO: Include this in the solve, rather than instantiating post-concretization
+    if "python" not in dependent_package.spec:
+        if extendee_spec:
+            python = extendee_spec
+        else:
```
Comment (scheibelp) [inline lib/spack/spack/solver/asp.py:4205]: The elif was removed on account of https://github.com/spack/spack/pull/34202/files#r1070191430  no longer applying.

**Review by tgamblin — APPROVED** [review]
This LGTM.  I think this is the right change to keep the current behavior until we can add proper dependencies for externals, but *not* expose it as part of the package API in 1.0.

**Comment by tgamblin** [comment]
@haampie if you like this one, can you approve and merge? I do not think we need to do a rebuild or anything with gitlab, as this logic only applies to externals.

**Review by haampie — APPROVED** [review]
